pkg/alertmanager/state_replication.go

```diff
@@ -115,13 +136,56 @@ func (s *state) Position() int {
 	return s.replicator.GetPositionForUser(s.userID)
 }
 
+// GetFullState returns the full internal state.
+func (s *state) GetFullState() (*clusterpb.FullState, error) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	all := &clusterpb.FullState{
+		Parts: make([]clusterpb.Part, 0, len(s.states)),
+	}
+
+	for key, s := range s.states {
+		b, err := s.MarshalBinary()
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to encode state for key: %v", key)
+		}
+		all.Parts = append(all.Parts, clusterpb.Part{Key: key, Data: b})
+	}
+
+	return all, nil
+}
+
 // starting waits until the alertmanagers are ready (and sets the appropriate internal state when it is).
 // The idea is that we don't want to start working" before we get a chance to know most of the notifications and/or silences.
 func (s *state) starting(ctx context.Context) error {
 	level.Info(s.logger).Log("msg", "Waiting for notification and silences to settle...")
 
-	// TODO: Make sure that the state is fully synchronised at this point.
+	// If the replication factor is <= 1, there is nowhere to obtain the state from.
+	if s.replicationFactor <= 1 {
+		level.Info(s.logger).Log("msg", "skipping settling (no replicas)")
+		return nil
+	}
+
 	// We can check other alertmanager(s) and explicitly ask them to propagate their state to us if available.
+	backoff := util.NewBackoff(ctx, s.settleBackoff)
```

Comment:
From a previous comment. Steve wrote:

> I suppose we want to try and cover transient errors as best as possible? But given we're using a more reliable means of fetching the state, which itself has a timeout, and as you say, we have another fallback to S3, maybe we can take the retries away entirely.

I'm leaning towards remove retries entirely, unless we have a good reason to keep it, because of the reasons already mentioned.

Reply:
@gotjosh

Reply:
The concept of retries in this scenario is more about making sure we've communicated with other replicas than it is about recovering from failures. 

What we care about here is two things:
- Make sure we've communicated with other replicas (if any)
- If no state was available from other replicas then try object storage

On scale up / downs retries might not matter all that much, but on total cluster failure or cluster startup I feel like we do care about retrying because other replicas might also starting up.

I guess the main thing to address as part of this logic IMO is that we should keep track of whenever `len(fullStates) == replication factor` and if is we should carry on.